src/controller/java/CHIPDeviceController-JNI.cpp

```diff
@@ -105,6 +105,79 @@ bool sShutdown      = false;
 jclass sAndroidChipStackCls              = NULL;
 jclass sChipDeviceControllerExceptionCls = NULL;
 
+// Callback for success and failure cases of GetConnectedDevice().
+struct GetConnectedDeviceCallback
+{
+    GetConnectedDeviceCallback(jobject javaCallback) :
+        mOnSuccess(OnDeviceConnectedFn, this), mOnFailure(OnDeviceConnectionFailureFn, this)
+    {
+        JNIEnv * env = chip::Controller::JniReferences::GetInstance().GetEnvForCurrentThread();
+        VerifyOrReturn(env != nullptr, ChipLogError(Controller, "Could not get JNIEnv for current thread"));
+        mJavaCallbackRef = env->NewGlobalRef(javaCallback);
+        if (mJavaCallbackRef == nullptr)
+        {
+            ChipLogError(Controller, "Could not create global reference for Java callback");
+        }
+    }
+
+    ~GetConnectedDeviceCallback()
+    {
+        JNIEnv * env = chip::Controller::JniReferences::GetInstance().GetEnvForCurrentThread();
+        VerifyOrReturn(env != nullptr, ChipLogError(Controller, "Could not get JNIEnv for current thread"));
+        env->DeleteGlobalRef(mJavaCallbackRef);
+    }
+
+    static void OnDeviceConnectedFn(void * context, Device * device)
+    {
+        JNIEnv * env         = JniReferences::GetInstance().GetEnvForCurrentThread();
+        auto * self          = static_cast<GetConnectedDeviceCallback *>(context);
+        jobject javaCallback = self->mJavaCallbackRef;
+
+        jclass getConnectedDeviceCallbackCls = nullptr;
+        JniReferences::GetInstance().GetClassRef(env, "chip/devicecontroller/GetConnectedDeviceCallback",
+                                                 getConnectedDeviceCallbackCls);
+        VerifyOrReturn(getConnectedDeviceCallbackCls != nullptr,
+                       ChipLogError(Controller, "Could not find GetConnectedDeviceCallback class"), delete self);
+
+        jmethodID successMethod;
+        JniReferences::GetInstance().FindMethod(env, javaCallback, "onDeviceConnected", "(J)V", &successMethod);
+        VerifyOrReturn(successMethod != nullptr, ChipLogError(Controller, "Could not find onDeviceConnected method"), delete self);
+
+        static_assert(sizeof(jlong) >= sizeof(void *), "Need to store a pointer in a Java handle");
+        env->CallVoidMethod(javaCallback, successMethod, reinterpret_cast<jlong>(device));
+        delete self;
```

Comment:
I'm pretty dubious of this.

What if the user backs out of the activity that needs to connect to the device? Probably we should drop the connection and cancel any callbacks at that point, and you can't do that because the callbacks own themselves.

One of the primary functions of the Callback class is to facilitate cancellation. For that to work, the activity needs to own the callbacks (most likely indirectly).

Reply:
I created `GetConnectedDeviceCallbackJni` which owns the underlying native callback. Added new files `AndroidCallbacks.h` and `AndroidCallbacks.cpp` to hold the `GetConnectedDeviceCallback` struct, and `AndroidCallbacks-JNI.cpp` to implement the new JNI functions.

I am still using finalizers here since the optimal solution I found for native cleanup isn't trivial to implement. Will clean up in a later pass.